me.id.webverify/webverifylib/src/main/java/me/id/webverifylib/RefreshAccessTokenHandler.java

```diff
@@ -0,0 +1,73 @@
+package me.id.webverifylib;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import me.id.webverifylib.listener.IDmeAccessTokenManagerListener;
+import me.id.webverifylib.listener.IDmeGetAccessTokenListener;
+import me.id.webverifylib.listener.IDmeScope;
+import me.id.webverifylib.networking.GetAccessTokenConnectionTask;
+
+/**
+ * Created by mirland on 23/02/17.
+ */
+final class RefreshAccessTokenHandler {
+  private static final Map<String, List<IDmeGetAccessTokenListener>> LISTENERS = new HashMap<>();
+  private static final Map<String, IDmeAccessTokenManagerListener> SCOPE_LISTENERS = new HashMap<>();
+
+  static void refreshAccessToken(IDmeScope scope, AuthToken authToken, IDmeGetAccessTokenListener listener) {
+    final String scopeId = authToken.getScopeId();
+    synchronized (LISTENERS) {
+      List<IDmeGetAccessTokenListener> scopeListeners = LISTENERS.get(scopeId);
+      if (scopeListeners == null) {
+        scopeListeners = new ArrayList<>();
+        LISTENERS.put(scopeId, scopeListeners);
+      }
+      if (scopeListeners.size() == 0) {
+        refreshToken(scope, authToken.getRefreshToken());
+      }
+      scopeListeners.add(listener);
+    }
+  }
+
+  private static synchronized IDmeAccessTokenManagerListener getScopeListener(IDmeScope scope) {
+    final String scopeId = scope.getScopeId();
+    IDmeAccessTokenManagerListener scopeListener = SCOPE_LISTENERS.get(scopeId);
```

Comment:
I think you can use a `ConcurrentMap` for this, taking advantage of the `putIfAbsent` method and avoiding the `synchronized`.